extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -590,51 +590,73 @@ export class Bind {
     });
   }
 
+
+  /**
+   * Determines which properties to update based on results of evaluation
+   * of all bound expression strings with the current scope. This method
+   * will only return properties that need to be updated along with their
+   * new value.
+   * @param {!Array<!BoundPropertyDef>} boundProperties
+   * @param {Object<string, ./bind-expression.BindExpressionResultDef>} results
+   * @return {
+   *   !Array<{
+   *     boundProperty: !BoundPropertyDef,
+   *     newValue: !./bind-expression.BindExpressionResultDef,
+   *   }>
+   * }
+   * @private
+   */
+  calculateUpdates_(boundProperties, results) {
+    const propertyUpdates = [];
+    boundProperties.forEach(boundProperty => {
+      const {expressionString, previousResult} = boundProperty;
+      const newValue = results[expressionString];
+      if (newValue === undefined ||
+          this.shallowEquals_(newValue, previousResult)) {
+        user().fine(TAG, `Expression result unchanged or missing: ` +
+            `"${expressionString}"`);
+        return;
+      }
+
+      boundProperty.previousResult = newValue;
+      user().fine(TAG, `New expression result: ` +
+          `"${expressionString}" -> ${newValue}`);
+      propertyUpdates.push({boundProperty, newValue});
+    });
+    return propertyUpdates;
+  }
+
   /**
    * Applies expression results to DOM.
    * @param {Object<string, ./bind-expression.BindExpressionResultDef>} results
    * @private
    */
   apply_(results) {
     const applyPromises = this.boundElements_.map(boundElement => {
+
+      let width, height;
       const {element, boundProperties} = boundElement;
+      const propertyUpdates = this.calculateUpdates_(boundProperties, results);
 
-      // TODO(choumx): We should avoid triggering a mutation if the expression
-      // results don't affect this element.
-      const applyPromise = this.resources_.mutateElement(element, () => {
+      if (propertyUpdates.length == 0) {
+        return Promise.resolve();
```

Comment:
Minor: Most of the elements in `applyPromises` will be resolved promises. In a follow-up PR, consider changing `this.boundElements_.map` to `forEach` and pushing mutation promises when necessary.